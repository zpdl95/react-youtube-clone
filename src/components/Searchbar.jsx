import React, { useEffect, useState, useRef } from 'react';
import { BsSearch } from 'react-icons/bs';
import { useNavigate, useParams } from 'react-router-dom';

export default function Searchbar() {
  const { query } = useParams();
  const [text, setText] = useState('');
  const navigate = useNavigate();
  const inputRef = useRef();

  const handleSubmit = (e) => {
    e.preventDefault();
    navigate(`/videos/${text}`);
    inputRef.current.blur();
  };

  useEffect(() => setText(query || ''), [query]);

  return (
    <form
      data-testid='search-form'
      className='w-full flex justify-center'
      onSubmit={handleSubmit}
    >
      <input
        className='w-7/12 p-2 outline-none bg-black'
        type='text'
        placeholder='Search'
        name='text'
        value={text}
        ref={inputRef}
        onChange={(e) => setText(e.target.value)}
      />
      <button className='bg-zinc-600 px-4'>
        <BsSearch />
      </button>
    </form>
  );
}
