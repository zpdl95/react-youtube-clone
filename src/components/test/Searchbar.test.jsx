import React from 'react';
import renderer from 'react-test-renderer';
import { Outlet, Route } from 'react-router-dom';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { withRouter } from '../../test/utils';
import Searchbar from '../Searchbar';

describe('Searchbar', () => {
  it('Searchbar snapshot', () => {
    const component = renderer.create(
      withRouter(<Route path='/' element={<Searchbar />} />)
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders with query correctly', () => {
    render(
      withRouter(<Route path='/:query' element={<Searchbar />} />, '/bts')
    );

    expect(screen.getByDisplayValue('bts')).toBeInTheDocument();
    expect(screen.getByRole('textbox')).toHaveValue('bts');
  });

  it('navigates to results page on searchbar submit', async () => {
    const query = 'fake-query';

    render(
      withRouter(
        <Route
          path='/'
          element={
            <>
              <Searchbar />
              <Outlet />
            </>
          }
        >
          <Route
            path={`/videos/${query}`}
            element={<p>{`Search result for ${query}`}</p>}
          />
        </Route>
      )
    );

    const searchInput = screen.getByRole('textbox');

    await userEvent.type(searchInput, query);
    await userEvent.type(searchInput, '{Enter}');
    // fireEvent.submit(screen.getByTestId('search-form'));

    // 버튼 방식
    // const searchButton = screen.getByRole('button');
    // const searchInput = screen.getByRole('textbox');
    // userEvent.type(searchInput, searchKeyword);
    // userEvent.click(searchButton);

    expect(screen.getByText(`Search result for ${query}`)).toBeInTheDocument();
  });
});
