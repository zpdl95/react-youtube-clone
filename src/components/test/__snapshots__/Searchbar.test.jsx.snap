// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Searchbar > Searchbar snapshot 1`] = `
<form
  className="w-full flex justify-center"
  data-testid="search-form"
  onSubmit={[Function]}
>
  <input
    className="w-7/12 p-2 outline-none bg-black"
    name="text"
    onChange={[Function]}
    placeholder="Search"
    type="text"
    value=""
  />
  <button
    className="bg-zinc-600 px-4"
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        {
          "color": undefined,
        }
      }
      viewBox="0 0 16 16"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M11.742 10.344a6.5 6.5 0 1 0-1.397 1.398h-.001c.03.04.062.078.098.115l3.85 3.85a1 1 0 0 0 1.415-1.414l-3.85-3.85a1.007 1.007 0 0 0-.115-.1zM12 6.5a5.5 5.5 0 1 1-11 0 5.5 5.5 0 0 1 11 0z"
      />
    </svg>
  </button>
</form>
`;
